### ISSUE TITLE
Issue accession serial only after the entry reaches a particular point in the workflow

### ISSUE THREAD

**Issue by brindakv** [informatics-isi-edu/protein-database#101]
This avoids gaps in the accession codes issued. Curators can still manually change the accession serial.

**Comment by svoinea** [comment]
The column `Accession_Serial` is of type `serial4` and it is initialized ONLY while creating a new record.
Currently, `chaise` does not allow columns of type `serial` to be editable.

**Comment by brindakv** [comment]
Yes, I realized that `Accession_Serial` is not editable by curators. Can it be initialized later in the workflow (e.g., after `Submit`) rather than while creating new record?

**Comment by svoinea** [comment]
No. `ermrest` does not support operations with `sequence`. It works only with `tables`. 
You can not get the value of a `sequence` through `ermrest` or set a value to it.
These operations can be done through the SQL client.

**Comment by hongsudt** [comment]
To avoid gap in Accession_Code, we will have to remove the Serial from the entry table and maintain it somewhere else. 
 
@svoinea what about creating a new table called "Accession_Code" or "Release_Entry" which contains: 
- Columns
  - Accession_Serial (Serial)
  - Accession_Code
  - Entry
- Keys: unique(Accession_Serial), unique(Accession_Code), unique(Entry)
- FKeys: Entry
- Note: we could create optional fk from entry.Accession_Code to Accession_Code.Accession_Code, but this might cause problems when an entry is deleted? 

A new entry is created only during the Pre-released process. 

With this method, we will remove the Accession_Serial from the entry table.

**Comment by svoinea** [comment]
My solution:
 - Create an `Accession_Serial` table `(Accession_Serial serial, Value text)`
 - The `entry.Accession_Serial` column type will be changed to `int` (therefore no initialization at create new record)
 - Whenever you want a new `Accession_Serial` value for the `entry` table, execute in `ermrest`: `INSERT INTO Accession_Serial(Value) VALUES ('dummy')`. It will return the row with the inserted values, therefore also the `Accession_Serial` value that will be set in the `entry` table. Then, you can delete the inserted record from the `Accession_Serial` table based on its `RID`.
 
In this way, the `Accession_Serial` table has just the role of generating a serial whenever it is "called" (through INSERT).

**Comment by svoinea** [comment]
Currently gaps of `Accession_Serial` in the `entry` table on production:
```
_ermrest_catalog_1=> select max("Accession_Serial") from entry;
 max 
-----
 158
(1 row)

_ermrest_catalog_1=> select * from generate_series(1,158) where generate_series not in (select "Accession_Serial" from entry) order by generate_series;
 generate_series 
-----------------
             100
             101
             102
             103
             104
             105
             106
             107
             108
             109
             110
             111
             118
             121
             122
             123
             127
             130
             131
             133
             136
             143
             148
             149
             150
             151
             153
             157
(28 rows)
```

**Comment by brindakv** [comment]
We will not back and fix the existing gaps in order to maintain chronological accession serials.

**Comment by svoinea** [comment]
Catalog `99` is ready for testing.

**Comment by brindakv** [comment]
@svoinea I cannot find the new `Accession_Code`table on `99` as user or curator. Does it require annotations to be set?

**Comment by brindakv** [comment]
Entry `D_1-XFEG` on `99` gives weird errors. Workflow was fine until `mmCIF Created`. At this point, I moved the entry to `Submission Complete` as curator. The expected behavior is that accession code will be issued and the workflow will move to `Hold`. Instead, I received an error email with `Release Ready` (which is not the current or the next workflow state), followed by another error email with `Submission complete` and a third one with an unexpected exception.

**Comment by brindakv** [comment]
- Define ACLs for the `Accession Code` table: Curator can view the table but not edit it; users cannot view it or edit it; in general no human intervention.
- Curators can move the workflow to `Submission Complete` to create accession code and move the workflow back to `Record Ready` in case accession codes need to be provided before completion of submission (note: errors need to be fixed though). @brindakv will test this (after bug is fixed). 
- Use `pdb-ihm` credentials for all automated scripts. 
- @svoinea will check the workflow bug.

**Comment by svoinea** [comment]
On catalog `99` I have:

 - removed the trigger before update the `Accession_Code` table
 - defined the `generated` annotation for the `Accession_Code` table such that nobody can edit it
 - defined ACLs such that submitters can view only the Accession_Code records corresponding to the entries they created

**Comment by brindakv** [comment]
@svoinea to review how the ACL for the `Accession_Code` changed such that it is now not seen publicly.

**Comment by brindakv** [comment]
> I cannot find the new Accession_Codetable on 99 as user or curator. Does it require annotations to be set?

Accession Code table can be seen from the `Data Categories` --> `Workflow` pulldown menu. 

> Entry D_1-XFEG on `99` gives weird errors. Workflow was fine until mmCIF Created. At this point, I moved the entry to Submission Complete as curator. The expected behavior is that accession code will be issued and the workflow will move to Hold. Instead, I received an error email with Release Ready (which is not the current or the next workflow state), followed by another error email with Submission complete and a third one with an unexpected exception.

This bug has been fixed. See new entry `D_1-XGTC` on catalog `99`. 

> Define ACLs for the Accession Code table: Curator can view the table but not edit it; users cannot view it or edit it; in general no human intervention.

Users can still see the Accession Code table but can only see their entries. Curator can view and `delete` all accession codes, but cannot edit. 

> Curators can move the workflow to Submission Complete to create accession code and move the workflow back to Record Ready in case accession codes need to be provided before completion of submission.

This works. 

> Use pdb-ihm credentials for all automated scripts.

This has been updated. 

> FKeys: Entry

[Ermrest](https://dev.pdb-dev.org/ermrest/catalog/99/schema/PDB/table/Accession_Code) shows that this foreign key points to `Entry.RID`. Not sure if this conform to the foreign key conventions described [here](https://docs.google.com/document/d/1lVgJIMEdmlQ7VKV_ZtshFhmvDVRS1HncsUOkItr_nyI/edit#heading=h.jzbsrji57fy3). 

> Create optional fk from entry.Accession_Code to Accession_Code.Accession_Code

Check whether this [foreign key](https://dev.pdb-dev.org/ermrest/catalog/99/schema/PDB/table/entry) conforms to the conventions.

**Comment by svoinea** [comment]
If the `entry.Accession_Code` is displayed as a column, then the curator will not have the link to navigate to the `Accession_Code` record.

If the `entry.Accession_Code` is displayed as a FK and the submitter has not at all access to the `Accession_Code` table, then he will not see the `entry.Accession_Code` value.

On catalog `99`, I have restricted the submitter to see only his records in the `Accession_Code` table w/o its `Accession_Serial` column.

**Comment by brindakv** [comment]
On catalog `99`, curators can see `Accession_Serial` but submitters cannot. @brindakv will verify.

**Comment by brindakv** [comment]
Looks good.

**Comment by svoinea** [comment]
The `dev` branch was merged, the new release `pdb-20221130.1` was created and deployed on production.
